src/firestore/delete.ts

```diff
@@ -0,0 +1,486 @@
+"use strict";
+
+import * as clc from "cli-color";
+import * as ProgressBar from "progress";
+
+import * as api from "../api";
+import * as firestore from "../gcp/firestore";
+import { FirebaseError } from "../error";
+import * as logger from "../logger";
+import * as utils from "../utils";
+
+// Datastore allowed numeric IDs where Firestore only allows strings. Numeric IDs are
+// exposed to Firestore as __idNUM__, so this is the lowest possible negative numeric
+// value expressed in that format.
+const MIN_ID = "__id-9223372036854775808__";
+
+export class FirestoreDelete {
+  /**
+   * Progress bar shared by the class.
+   */
+  static progressBar = new ProgressBar("Deleted :current docs (:rate docs/s)\n", {
+    total: Number.MAX_SAFE_INTEGER,
+  });
+
+  public isDocumentPath: boolean;
+  public isCollectionPath: boolean;
+  public path: string;
+
+  private project: string;
+  private recursive: boolean;
+  private shallow: boolean;
+  private allCollections: boolean;
+
+  private allDescendants: boolean;
+  private root: string;
+  private parent: string;
+
+  /**
+   * Construct a new Firestore delete operation.
+   *
+   * @constructor
+   * @param {string} project the Firestore project ID.
+   * @param {string | undefined} path path to a document or collection.
+   * @param {boolean} options.recursive true if the delete should be recursive.
+   * @param {boolean} options.shallow true if the delete should be shallow (non-recursive).
+   * @param {boolean} options.allCollections true if the delete should universally remove all collections and docs.
```

Comment:
All the types can be dropped since you now type them in the function signature! Applies throughout.